src/logger.cr

```diff
@@ -33,6 +33,14 @@ module Crylog
 
     # Urgent alert.
     Emergency = 600
+
+    # :nodoc:
+    @@upcase_pool = Hash(Severity, String).new
+
+    # Returns memoized upcased String.  Used in Formatter.
+    def upcase_to_s
+      @@upcase_pool[self] ||= self.to_s.upcase
```

Comment:
Already tried, but sadly the use case and API aren't compatible.  `StringPool` is close to a `Set` of Strings.  `upcase_to_s` maps the `Enum` to a fixed `String`.

Reply:
Ah right.  Since we know all the strings, would it be worth using a case statement and just do like
```cr
case self
when .debug? then "DEBUG"
...
end
```

Reply:
@didactic-drunk Thoughts on that ^?

Reply:
Do you know the speed difference between the `case` and a `Hash` lookup?  I don't.  Is a benchmark needed?

A`to_s` override works either way way. I thought maybe it was bad form to override `#to_s` on an `enum`.  It doesn't look like upcase would harm `Enum#parse`.

Reply:
Case is much much faster

```cr
enum Severity1
  Debug  = 100
  Info   = 200
  Notice = 250

  @@upcase_pool = Hash(Severity1, String).new

  def upcase_to_s
    @@upcase_pool[self] ||= self.to_s.upcase
  end
end

enum Severity2
  Debug  = 100
  Info   = 200
  Notice = 250

  def to_s
    case self
    when .debug?  then "DEBUG"
    when .info?   then "INFO"
    when .notice? then "NOTICE"
    end
  end
end

require "benchmark"

Benchmark.ips do |x|
  x.report("hash") do
    Severity1::Debug.upcase_to_s
    Severity1::Info.upcase_to_s
    Severity1::Notice.upcase_to_s
  end

  x.report("case") do
    Severity2::Debug.to_s
    Severity2::Info.to_s
    Severity2::Notice.to_s
  end
end


hash  35.55M ( 28.13ns) (± 0.62%)  0.0B/op  18.05× slower
case 641.84M (  1.56ns) (± 1.04%)  0.0B/op        fastest
```

Reply:
I had no idea.  Would you like it as a each `Enum` macro?

Reply:
Naw, just manually making the case is prob good enough.  These string aren't going to change so no need to make it dynamic.